src/test/testUtils.ts

```diff
@@ -73,3 +73,141 @@ export const StepProgramGiveClue = async (
+
+  while (numbers.size < 4) {
+    numbers.add(generateRandomNumber());
+  }
+
+  return Array.from(numbers);
+}
+
+/**
+ * In this function, guesses are obtained from a predefined list of actions.
+ * @param winnerFlag - A flag that determines the outcome of the function: `0` produces results on behalf of the codeMaster, `1` for the codeBreaker, and `2` for an unsolved game case.
+ * @param actions - A predetermined list of number sequences that can represent guesses or solutions. These are used to simulate guesses.
+ * @param round - The number of rounds to be simulated. In each round, both `.makeGuess()` and `.giveClue()` methods are executed; hence, one round consists of two consecutive turns.
+ * @param lastProof - The `StepProgramProof` instance that initiates the recursive proof generation process.
+ * @param codeBreakerKey - The private key of the codeBreaker, used for signature creation.
+ * @param codeMasterKey - The private key of the codeMaster, used for signature creation.
+ * @param salt - The salt used in the hash function to prevent pre-image attacks.
+ * @param secretCombination - The secret combination of the game, specified by the codeMaster.
+ * @returns - A `StepProgramProof` instance.
+ */
+export const generateTestProofs = async (
+  round: number,
+  winnerFlag: number,
+  actions: typeof guessConfig1,
+  lastProof: StepProgramProof,
+  codeBreakerKey: PrivateKey,
+  codeMasterKey: PrivateKey,
+  salt: Field
+): Promise<StepProgramProof> => {
+  const secretCombination = actions.secret;
+  const guesses = actions.totalAttempts;
+
+  if (winnerFlag == 0) {
+    for (let i = 0; i < round; i++) {
+      const guess = guesses[i];
+
+      lastProof = await StepProgramMakeGuess(lastProof, guess, codeBreakerKey);
+
+      lastProof = await StepProgramGiveClue(
+        lastProof,
+        secretCombination,
+        salt,
+        codeMasterKey
+      );
+    }
+
+    return lastProof;
+  } else if (winnerFlag == 1) {
+    for (let i = 0; i < round - 1; i++) {
+      const guess = guesses[i];
+
+      lastProof = await StepProgramMakeGuess(lastProof, guess, codeBreakerKey);
+
+      lastProof = await StepProgramGiveClue(
+        lastProof,
+        secretCombination,
+        salt,
+        codeMasterKey
+      );
+    }
+
+    lastProof = await StepProgramMakeGuess(
+      lastProof,
+      secretCombination,
+      codeBreakerKey
+    );
+
+    lastProof = await StepProgramGiveClue(
+      lastProof,
+      secretCombination,
+      salt,
+      codeMasterKey
+    );
+
+    return lastProof;
+  } else if (winnerFlag == 2) {
+    for (let i = 0; i < round; i++) {
+      let guess = generateRandomGuess();
+
+      lastProof = await StepProgramMakeGuess(lastProof, guess, codeBreakerKey);
+
+      lastProof = await StepProgramGiveClue(
+        lastProof,
+        secretCombination,
+        salt,
+        codeMasterKey
+      );
+    }
+    return lastProof;
+  } else {
+    throw new Error('WinnerFlag is not valid!');
+  }
```

Comment:
```suggestion
  for (let i = 0; i < (winnerFlag === 1 ? round - 1 : round); i++) {
    const guess = winnerFlag === 2 ? generateRandomGuess() : guesses[i];

    lastProof = await StepProgramMakeGuess(lastProof, guess, codeBreakerKey);

    lastProof = await StepProgramGiveClue(
      lastProof,
      secretCombination,
      salt,
      codeMasterKey
    );
  }

  if (winnerFlag === 1) {
    lastProof = await StepProgramMakeGuess(
      lastProof,
      secretCombination,
      codeBreakerKey
    );

    lastProof = await StepProgramGiveClue(
      lastProof,
      secretCombination,
      salt,
      codeMasterKey
    );
  }

  return lastProof;
  ```
  We can simplify that part like this to reduce repetitive code blocks